cliquet/permission/postgresql/__init__.py

```diff
@@ -0,0 +1,178 @@
+           AND principal = %(principal)s;"""
+        with self.connect() as cursor:
+            cursor.execute(query, dict(user_id=user_id, principal=principal))
+
+    def user_principals(self, user_id):
+        query = """
+        SELECT principal
+          FROM user_principals
+         WHERE user_id = %(user_id)s;"""
+        with self.connect() as cursor:
+            cursor.execute(query, dict(user_id=user_id))
+            results = cursor.fetchall()
+        return set([r['principal'] for r in results])
+
+    def add_principal_to_ace(self, object_id, permission, principal):
+        query = """
+        INSERT INTO aces (object_id, permission, principal)
+        SELECT %(object_id)s, %(permission)s, %(principal)s
+         WHERE NOT EXISTS (
+            SELECT principal
+              FROM aces
+             WHERE object_id = %(object_id)s
+               AND permission = %(permission)s
+               AND principal = %(principal)s
+        );"""
+        with self.connect() as cursor:
+            cursor.execute(query, dict(object_id=object_id,
+                                       permission=permission,
+                                       principal=principal))
+
+    def remove_principal_from_ace(self, object_id, permission, principal):
+        query = """
+        DELETE FROM aces
+         WHERE object_id = %(object_id)s
+           AND permission = %(permission)s
+           AND principal = %(principal)s;"""
+        with self.connect() as cursor:
+            cursor.execute(query, dict(object_id=object_id,
+                                       permission=permission,
+                                       principal=principal))
+
+    def object_permission_principals(self, object_id, permission):
+        query = """
+        SELECT principal
+          FROM aces
+         WHERE object_id = %(object_id)s
+           AND permission = %(permission)s;"""
+        with self.connect() as cursor:
+            cursor.execute(query, dict(object_id=object_id,
+                                       permission=permission))
+            results = cursor.fetchall()
+        return set([r['principal'] for r in results])
+
+    def object_permission_authorized_principals(self, object_id, permission,
+                                                get_bound_permissions=None):
+        if get_bound_permissions is None:
+            perms = [(object_id, permission)]
+        else:
+            perms = get_bound_permissions(object_id, permission)
+
+        perms = ','.join(["('%s', '%s')" % p for p in perms])
+        query = """
+        WITH required_perms AS (
+          VALUES %s
+        )
+        SELECT principal
+          FROM required_perms JOIN aces
+            ON (object_id = column1 AND permission = column2)
+        ;""" % perms
+        with self.connect() as cursor:
+            cursor.execute(query)
+            results = cursor.fetchall()
+        return set([r['principal'] for r in results])
+
```

Comment:
Ah, ok. I just read the code of ``BasePermission.check_permission``.

But then why do we need ``object_permission_authorized_principals`` ?